src/org/nfrac/comportex/cells.cljc

```diff
@@ -735,58 +737,97 @@
                                                     (:distal-bits distal-state)
                                                     (:distal-exc distal-state)
                                                     (:seg-learn-threshold spec)
                                                     (:depth spec))
                             (merge-with + tp-exc))
           ;; find active and winner cells in the columns
+          pc (:pred-cells distal-state)
+          depth (:depth spec)
+          prior-col-winners (:col-winners state)
           [rng* rng] (random/split rng)
           {ac :active-cells
-           lc :winner-cells
-           b-cols :burst-cols
+           col-winners :col-winners
+           burst-cols :burst-cols
            stable-ac :stable-active-cells}
-          (select-active-cells a-cols rel-cell-exc (:pred-cells distal-state) spec rng*)
+          (select-active-cells a-cols rel-cell-exc
+                               ;; definition of bursting for a column
+                               (fn [col win-cell col-ac]
+                                 (if (and (not newly-engaged?)
+                                          (= win-cell (prior-col-winners col)))
+                                   ;; for continuing temporal pooling
+                                   (== depth (count col-ac))
+                                   ;; otherwise: for discrete transitions
+                                   (not (or (pc win-cell) (tp-exc win-cell)))))
+                               (:col-winners state) ;; keep winners stable
+                               spec rng*)
+          ;; learning cells are the winning cells, but excluding any
+          ;; continuing winners when temporal pooling
+          old-winners (vals (:col-winners state))
+          new-winners (vals col-winners)
+          learning (if newly-engaged? ;; always true at first level
+                     new-winners
+                     (remove (set old-winners) new-winners))
           ;; update continuing TP activation
-          lc-stable-exc (for [cell-id lc
-                              :let [[col ci] cell-id
-                                    exc (stable-col-exc [col 0])]
-                              :when exc]
-                          [cell-id exc])
-          next-tp-exc (-> (apply-excitation tp-exc
-                                            lc-stable-exc
-                                            (:temporal-pooling-max-exc spec)
-                                            (:temporal-pooling-amp spec))
-                          ;; clear TP excitation for inhibited cells
-                          (select-keys lc))
-          depth (:depth spec)]
+          next-tp-exc (if higher-level?
+                        (let [new-ac (if newly-engaged?
+                                       ac
+                                       (set/difference ac (:active-cells state)))]
+                          (into tp-exc
```

Comment:
Trying to piece this together...

- If the layer is not engaged, then all newly active cells get their tp-exc set to the max. Those that were already active don't, so they decay in each timestep.
- If the layer is newly engaged, all active cells get their tp-exc set to the max.
- If the layer is engaged (but not newly), then all newly active cells get their tp-exc set to the max. The others decay in each timestep (because of the commented out `true ;(not engaged?)`)

Am I reading it right? I haven't grokked the change yet, I'm expecting it to all crystallize after I sleep on it.